Msi 2083 create in store pickup shipping method

swnsma opened the pull request:
### Description (*)
Add In-Store Pickup Delivery method.
Please proceed after #2214.
Please check my notices.

### Fixed Issues (if relevant)
1. magento-engcom/msi#2083: Create 'In Store Pickup' Shipping Method.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Move Pickup Locations Mapping to global area.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Add In-Store Pickup Delivery method.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Adjust old tests and tests coverage for In-Store Pickup Delivery Method.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Remove not used configuration.

swnsma commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetIsAnyPickupLocationAvailable.php:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Check if any pickup location is available in current scope.
+ */
+class GetIsAnyPickupLocationAvailable
```
Reformat and create as SPI?

swnsma commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetIsAnyPickupLocationAvailable.php:
```diff
@@ -0,0 +1,96 @@
+     *
+     * @var array
+     */
+    private $isAvailable = [];
```
Runtime caching.
Seems strange that command has state but added due to multiple execution of service.

swnsma commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/InStorePickup.php:
```diff
@@ -0,0 +1,186 @@
+/**
+ * In-Store Pickup Delivery Method Model.
+ */
+class InStorePickup extends AbstractCarrier implements CarrierInterface
```
Probably, need to be moved to Magento_InventoryInStorePickupApi.

ishakhsuvarov replied:
Let's find a way to avoid all of these dependencies in API

swnsma commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/InStorePickup.php:
```diff
@@ -0,0 +1,186 @@
+class InStorePickup extends AbstractCarrier implements CarrierInterface
+{
+    private const CARRIER_CODE = 'in_store';
+    private const METHOD_CODE = 'pickup';
```
Naming to have 'in_store_pickup' method name in result.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Refactoring.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Add missed dependencies.

swnsma commented:
I do not agree with Static Tests as module uses classes from **Magento\Config, Magento\Shipping** in system.xml.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Refactoring.

ishakhsuvarov reviewed (commented):
Do you mind if we work on this one once #2214 is merged?

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetFreeBoxes.php:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Calculate and return number of items with free delivery.
+ */
+class GetFreeBoxes
```
Boxes? Maybe packages or something?

swnsma replied:
Renamed to Packages.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetFreeBoxes.php:
```diff
@@ -0,0 +1,56 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickup\Model\Carrier\Command;
```
Do we really want shipping method right inside the `InStorePickup` module?

swnsma replied:
Moved to separated module.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetFreeBoxes.php:
```diff
@@ -0,0 +1,56 @@
+        if ($request->getAllItems()) {
+            /** @var \Magento\Quote\Model\Quote\Item\AbstractItem $item */
+            foreach ($request->getAllItems() as $item) {
+                if ($item->getProduct()->isVirtual() || $item->getParentItem()) {
```
Would be great to avoid breaking the Law of Demeter if those older APIs would allow it.

swnsma replied:
I am not sure if it was a good idea, because we create new dependency on `Magento_Catalog` but please check changes in my 1d903059b71201bb607fa36975083a46ee37d982.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetIsAnyPickupLocationAvailable.php:
```diff
@@ -0,0 +1,96 @@
+     *
+     * @return bool
+     */
+    public function execute(int $websiteId): bool
```
It'd be better to rely on sales channel. Someone who puts a lot of effort into implementing any channel other then website would be grateful for your effort.

swnsma replied:
Done.
Probably, it can be exposed as API endpoint.

swnsma replied:
By the way, does it will be more logical for all of these services to use the same root:
```
\Magento\InventoryInStorePickupApi\Api\GetNearbyPickupLocationsInterface
\Magento\InventoryInStorePickupApi\Api\GetPickupLocationsAssignedToStockOrderedByPriorityInterface
\Magento\InventoryInStorePickupApi\Api\GetIsAnyPickupLocationAvailableInterface
```

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/Command/GetShippingPrice.php:
```diff
@@ -0,0 +1,26 @@
+namespace Magento\InventoryInStorePickup\Model\Carrier\Command;
+
+use Magento\InventoryInStorePickupApi\Model\Carrier\GetShippingPriceInterface;
+use Magento\Quote\Model\Quote\Address\RateRequest;
```
One more evidence that we need a separate module. I;d suggest not to depend on quote here.

swnsma replied:
Moved to separated module.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/InStorePickup.php:
```diff
@@ -0,0 +1,178 @@
+    private $getFreeBoxes;
+
+    /**
+     * InStorePickup constructor.
```
👍

swnsma replied:
Eliminated.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/Carrier/InStorePickup.php:
```diff
@@ -0,0 +1,178 @@
+
+        $shippingPrice = $this->getFinalPriceWithHandlingFee($shippingPrice);
+
+        if ($shippingPrice !== null && $request->getPackageQty() == $freeBoxes) {
```
let's try to stick to strict comparison

swnsma replied:
Done.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/etc/config.xml:
```diff
@@ -22,5 +22,17 @@
+                <sallowspecific>0</sallowspecific>
+                <model>Magento\InventoryInStorePickup\Model\Carrier\InStorePickup</model>
+                <name>Pick in Store</name>
+                <title>In-Store Delivery</title>
```
Delivery?

swnsma replied:
In-Store Pickup Delivery method with Pick in Store rate.
I will be happy to see other variants :)

ishakhsuvarov reviewed (changes requested):
My main concern for now is `Quote` dependencies.
Additionally, what if someone would want to build another shipping method for store pickup? That's a lot of customization for the main pickup module.

Commit:
Merge branch 'store-pickup' into MSI-2083-Create-In-Store-Pickup-Shipping-Method

Commit:
Merge remote-tracking branch 'origin/store-pickup' into MSI-2083-Create-In-Store-Pickup-Shipping-Method

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Fix broken dependency.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Refactoring of In-Store Pickup Delivery Method.
Provide extension/customization points for:
- Validation of Rate Request for In-Store Pickup Delivery
- Free packages calculation
- Creation of Shipping Price Request
- In-Store Pickup Delivery Price Calculation.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Avoid breaking Law of Demeter.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Fix failed static tests.

swnsma commented:
Please continue with review.
Static tests failed due to they do not check dependencies in system.xml.

ishakhsuvarov commented:
@swnsma Working on it. Going to take some time looking at PR size :)

ishakhsuvarov committed:
Merge branch 'store-pickup' into MSI-2083-Create-In-Store-Pickup-Shipping-Method

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Rename Checkout -> Shipping.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Remove interfaces for price calculation and free packages count.

Commit:
MSI-2083 Create 'In Store Pickup' Shipping Method.
Refactoring.
- fix failed tests
- change Localized exception to InvalidArgumentException for Shipping Request Validator Chain
- add new line to EOL

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Model/GetIsAnyPickupLocationAvailable.php:
```diff
@@ -0,0 +1,71 @@
+     */
+    public function execute(SalesChannelInterface $salesChannel): bool
+    {
+        return $this->checkPickupLocationsAvailability($salesChannel);
```
Why do you proxy the call to another method?

swnsma replied:
Refactored.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupShippingApi/Model/Carrier/Validation/RequestValidatorChain.php:
```diff
@@ -0,0 +1,73 @@
+
+        foreach ($validators as $validator) {
+            if (!$validator instanceof RequestValidatorInterface) {
+                throw new LocalizedException(
```
I'd change it to `InvalidArgumentException`, this kind of error does not make sense to customers anyway.

swnsma replied:
Done.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickupShippingApi/Model/Carrier/Validation/RequestValidatorChain.php:
```diff
@@ -0,0 +1,73 @@
+            if (!$validator instanceof RequestValidatorInterface) {
+                throw new LocalizedException(
+                    __(
+                        'In-Store Pickup Carrier Rate Request Validator must implement %1.',
```
If we decide not to use `Localized` then translation may be skipped. Also, message may be made more informative my providing not only the expected interface but the class which triggered error. Something like `getclass($validator)`.

swnsma replied:
Agree.

ishakhsuvarov reviewed (changes requested):
Overall, looks really good to me.

ishakhsuvarov commented on app/code/Magento/InventoryInStorePickup/Test/_files/place_order.php:
```diff
@@ -25,13 +23,4 @@
     ->create();
 $cart = current($cartRepository->getList($searchCriteria)->getItems());
 
-$product = $productRepository->get('SKU-1');
```
I believe some of the older tests got affected by this change as they can't place order with empty cart.

swnsma replied:
Fixed.
Will check with Travis results.

swnsma commented:
@ishakhsuvarov please continue review.

ishakhsuvarov commented:
@swnsma Here we go again

ishakhsuvarov committed:
magento-engcom/msi#2224: Store Pickup Shipping Method
 - Make dependency test happy

ghost commented:
Hi @swnsma, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento-engcom+msi+2224), it will take less than a minute. </br>Your feedback will help us to improve contribution process.
